avr-hal-generic/src/serial.rs

```diff
@@ -1,5 +1,75 @@
 //! Serial Implementations
 
+use core::cmp::Ordering;
+
+// Clock is needed because the calculations needs to take core clock into account
+#[derive(Debug, Clone, Copy)]
+pub struct Baudrate<CLOCK> {
+    pub ubrr: u16,
+    pub u2x: bool,
+    pub _clock: ::core::marker::PhantomData<CLOCK>,
+}
+
+impl<CLOCK: crate::clock::Clock> PartialEq for Baudrate<CLOCK> {
+    fn eq(&self, other: &Self) -> bool {
+        self.value() == other.value()
+    }
+}
+
+impl<CLOCK: crate::clock::Clock> Eq for Baudrate<CLOCK> {}
+
+impl<CLOCK: crate::clock::Clock> PartialOrd for Baudrate<CLOCK> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.value().cmp(&other.value()))
+    }
+}
+
+impl<CLOCK: crate::clock::Clock> Ord for Baudrate<CLOCK> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.value().cmp(&other.value())
+    }
+}
+
+impl<CLOCK: crate::clock::Clock> Baudrate<CLOCK> {
+    pub fn new(baud: u32) -> Baudrate<CLOCK> {
+        // Try u2x mode first
+        let mut baud_setting = (CLOCK::FREQ / 4 / baud - 1) / 2;
+        let mut u2x = true;
+
+        // Switch back from u2x if required
+        if ((CLOCK::FREQ == 16_000_000) && (baud == 57_600)) || (baud_setting > 4095) {
+            u2x = false;
+            baud_setting = (CLOCK::FREQ / 8 / baud - 1) / 2;
+        }
+
+        let ubrrh = baud_setting >> 8;
+        let ubrrl = baud_setting;
+
+        Baudrate {
+            ubrr: (ubrrh | ubrrl) as u16,
```

Comment:
You messed something up here.  I think you copied this from the arduino-core but over there something different is done:  Writing to the `*_ubrrh` and `*_ubrrl` variables _is_ writing to the ubrr register high and low bytes.  Your `baud_setting` already contains the complete 16-bit ubrr value so that's what you should be storing:

```suggestion
        debug_assert!(baud_setting <= u16::MAX);

        Baudrate {
            ubrr: baud_setting as u16,
```